src/google/protobuf/compiler/objectivec/objectivec_file.cc

```diff
@@ -239,16 +247,26 @@ void FileGenerator::GenerateHeader(io::Printer* printer) {
       "\n",
       "google_protobuf_objc_version", StrCat(GOOGLE_PROTOBUF_OBJC_VERSION));
 
-  // #import any headers for "public imports" in the proto file.
+  // The bundled protos (WKTs) force off the use of forward declarations.
+  bool headers_use_forward_declarations =
+      generation_options_.headers_use_forward_declarations && !is_bundled_proto_;
+
   {
     ImportWriter import_writer(
         generation_options_.generate_for_named_framework,
         generation_options_.named_framework_to_proto_path_mappings_path,
         generation_options_.runtime_import_prefix,
         /* include_wkt_imports = */ false);
     const std::string header_extension(kHeaderExtension);
-    for (int i = 0; i < file_->public_dependency_count(); i++) {
-      import_writer.AddFile(file_->public_dependency(i), header_extension);
+    if (headers_use_forward_declarations) {
+      // #import any headers for "public imports" in the proto file.
+      for (int i = 0; i < file_->public_dependency_count(); i++) {
+        import_writer.AddFile(file_->public_dependency(i), header_extension);
+      }
+    } else {
+      for (int i = 0; i < file_->dependency_count(); i++) {
```

Comment:
So there are some edge cases here.

If the proto import is only for an enum and the enum is the value of a repeated field or a map, then we don't really need the types because there isn't covariance for those.

Similarly, if the imported types are for extensions, then the ObjC api doesn't need the types here, as they don't end up in the generated code.

So this code path could be much more complex and figure out what headers are really needed and only import those, but the complexity doesn't seem worth it.

Reply:
agreed.